ptbcontrib/persistent_jobstore/jobstore.py

```diff
@@ -0,0 +1,107 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser Public License for more details.
+#
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This file contains AdaptedSQLAlchemyJobStore."""
+
+from typing import Any
+import pickle
+import telegram
+import apscheduler.job
+
+from apscheduler.jobstores.sqlalchemy import SQLAlchemyJobStore
+from telegram.ext import CallbackContext, Dispatcher
+
+
+class AdaptedSQLAlchemyJobStore(SQLAlchemyJobStore):
+    """
+    Wraps apscheduler.SQLAlchemyJobStore to make telegram Job class storable.
+    """
+
+    def __init__(self, dispatcher: Dispatcher, *args: Any, **kwargs: Any) -> None:
+        """
+        :param dispatcher (:class:`telegram.ext.Dispatcher`): Dispatcher instance
+            that will be passed to CallbackContext when recreating jobs.
+        :param args: Arguments to be passed to the SQLAlchemyJobStore constructor.
+        """
+
+        super().__init__(*args, **kwargs)
+        self.dispatcher = dispatcher
+
+    def add_job(self, job: apscheduler.job) -> None:
+        """
+        Called from apscheduler's internals after adding a new job.
+        Args:
+            job (:obj:`apscheduler.job`): The job to be persisted.
+        """
+
+        job = self._prepare_job(job)
+        super().add_job(job)
+
+    def update_job(self, job: apscheduler.job) -> None:
+        """
+        Called from apscheduler's internals after updating a job.
+        Args:
+            job (:obj:`apscheduler.job`): The job to be updated.
+        """
+        job = self._prepare_job(job)
+        super().update_job(job)
+
+    @staticmethod
+    def _prepare_job(job: apscheduler.job) -> apscheduler.job:
+        """
+        Erase all unpickable data from telegram.ext.Job
+        Args:
+            job (:obj:`apscheduler.job`): The job to be processed.
+        """
+
+        # while updating existing job, job.args[0]
+        # will be a dict fetched from db instead
+        # of CallbackContext.
+        if isinstance(job.args[0], dict):
+            job_elements = job.args[0]
+        else:
+            # CallbackContext.job
+            tg_job = job.args[0].job
+            job_elements = {"context": tg_job.context, "name": tg_job.name}
+
+        job.args = (job_elements,)  # ensure tuple by (,)
+        return job
+
+    def _reconstitute_job(self, job_state: bytes) -> apscheduler.job:
+        """
+        Called from apscheduler's internals when loading job.
+        Args:
+            job_state (:obj:`str`): String containing pickled job state.
+        """
+
+        job_state_dict = pickle.loads(job_state)
+        telegram_job_elements = job_state_dict["args"][0]
+        tg_job = telegram.ext.Job(
+            callback=None,
```

Comment:
why is the callback `None`? how do we get the actual callback back?

Reply:
When job is first added it's stored in database along with callback, here we are only modifying "arg" of the job and using `tg_job` to build `CallbackContext`.

Reply:
Ah, okay 👍 . worth a comment, maybe. Btw, this only works with context-based callbacks, right? Not a problem, but should probably be noted in the readme

Reply:
Alright!